### PR TITLE
Restructure CLI Commands

### LINKED ISSUE #8756: Proposal: Group available commands in the command line help text into "general", "image centric" and "container centric"
Newbie here, who had a hard time "getting" how to work with docker, until I realized there are 2-3 categories of commands available to the docker command: General commands, image centric ones, and container centric ones (and I had to look up the difference between the last two as well).

Currently, all the commands are mixed together in a long and thus comparably unreadable list:

````
[18:35:21] ~ $ docker
Usage: docker [OPTIONS] COMMAND [arg...]
 -H=[unix:///var/run/docker.sock]: tcp://host:port to bind/connect to or unix://path/to/socket to use

A self-sufficient runtime for linux containers.

Commands:
    attach    Attach to a running container
    build     Build an image from a Dockerfile
    commit    Create a new image from a container's changes
    cp        Copy files/folders from the containers filesystem to the host path
    diff      Inspect changes on a container's filesystem
    events    Get real time events from the server
    export    Stream the contents of a container as a tar archive
    history   Show the history of an image
    images    List images
    import    Create a new filesystem image from the contents of a tarball
    info      Display system-wide information
    inspect   Return low-level information on a container
    kill      Kill a running container
    load      Load an image from a tar archive
    login     Register or Login to the docker registry server
    logs      Fetch the logs of a container
    port      Lookup the public-facing port which is NAT-ed to PRIVATE_PORT
    pause     Pause all processes within a container
    ps        List containers
    pull      Pull an image or a repository from the docker registry server
    push      Push an image or a repository to the docker registry server
    restart   Restart a running container
    rm        Remove one or more containers
    rmi       Remove one or more images
    run       Run a command in a new container
    save      Save an image to a tar archive
    search    Search for an image in the docker index
    start     Start a stopped container
    stop      Stop a running container
    tag       Tag an image into a repository
    top       Lookup the running processes of a container
    unpause   Unpause a paused container
    version   Show the docker version information
    wait      Block until a container stops, then print its exit code
````

I think, grouping them into the categories methined above ("general", "image centric" and "container centric") would go a very long way in making this list less confusing, and to highlight the difference between images and containers, from the start, to help new users note the difference. 

So, what about grouping them something like this (Not sure I've classified the commands 100% correctly, so correct me on that):

````
[18:35:21] ~ $ docker
Usage: docker [OPTIONS] COMMAND [arg...]
 -H=[unix:///var/run/docker.sock]: tcp://host:port to bind/connect to or unix://path/to/socket to use

A self-sufficient runtime for linux containers.

General commands:
    events    Get real time events from the server
    info      Display system-wide information
    login     Register or Login to the docker registry server
    version   Show the docker version information

Commands for working with images:
    build     Build an image from a Dockerfile
    history   Show the history of an image
    images    List images
    import    Create a new filesystem image from the contents of a tarball
    load      Load an image from a tar archive
    pull      Pull an image or a repository from the docker registry server
    push      Push an image or a repository to the docker registry server
    rmi       Remove one or more images
    save      Save an image to a tar archive
    search    Search for an image in the docker index
    tag       Tag an image into a repository

Commands for working with containers (running instances of images):
    attach    Attach to a running container
    commit    Create a new image from a container's changes
    diff      Inspect changes on a container's filesystem
    cp        Copy files/folders from the containers filesystem to the host path
    export    Stream the contents of a container as a tar archive
    inspect   Return low-level information on a container
    kill      Kill a running container
    logs      Fetch the logs of a container
    port      Lookup the public-facing port which is NAT-ed to PRIVATE_PORT
    pause     Pause all processes within a container
    ps        List containers
    ps -a     List all containers (including stopped ones)
    restart   Restart a running container
    rm        Remove one or more containers
    run       Run a command in a new container
    start     Start a stopped container
    stop      Stop a running container
    top       Lookup the running processes of a container
    unpause   Unpause a paused container
    wait      Block until a container stops, then print its exit code
````

(I also added `ps -a` here, since it was such a crucial revelation for me when I realized that is what I need in order to see containers that I had just ran, but suddently couldn't find)

Wadda'ya think?

### LINKED ISSUE #8829: Preserve / improve consistency in CLI UI
**Summary**: 
There is no standardised "naming" for (sub)commands on the CLI (e.g. plural vs singular, create vs add), sometimes leading to inconsistencies.  

**Proposed solution**:
A design document should be written, describing the (naming) conventions to follow for commands and the UX in general.

----

A number of new features are currently proposed via PRs for addition to Docker. These features add new commands to the CLI and I have noticed that the UI for those commands (and *existing* commands) is inconsistent.

While there still a chance to change those proposals before they are merged, I think it would be a good thing to review those PRs and reach consensus on what the UI should look like to be consistent.

I've created a (far from complete) overview of some *existing* and *proposed* commands;

Entity | Action | Command | Notes
--- | --- | --- | ---
Container | list | `docker ps`
Device | list | - | (not supported)
Host | list | `docker hosts` | no sub command. see https://github.com/docker/docker/issues/8681
Image | list | `docker images` | no sub command (`ls`)
Links | list | `docker links list` | see https://github.com/docker/docker/issues/7468
Volume | list | `docker volumes ls` | see https://github.com/docker/docker/pull/8484
Container | delete | `docker rm (id)`
Device | delete | `docker device remove (id)` | see https://github.com/docker/docker/pull/8826
Host | delete | - | (not supported?)
Image | delete | `docker rmi (id)`
Links | delete | `docker links remove (id)`
Volume | delete | `docker volumes rm (id)`
Container | create | `docker create` | no "containers" before "create"
Device | create | `docker device add` | note: *singular* "device", others are "plural"
Host | create | `docker hosts create` | `run` [is proposed as alternative](https://github.com/docker/docker/issues/8681#issuecomment-60349475)
Link | create | `docker links add` |
Volume | create | `docker volumes add`

- The `ls` and `list` subcommands could be omitted dropped for `volumes` and `links` to be in line with the existing `images` command. 
- The subcommand for "delete" should either be "remove" or "rm" for all entities.
- The subcommand for "create" could be renamed from "add" to "create" to be in line with the existing `docker create` command (not sure, I actually like "add")
- All entities / top-level commands should be either plural or singular (rename `device` to `devices`?)

Now that the "links" (https://github.com/docker/docker/issues/7468), "volumes" (https://github.com/docker/docker/pull/8484), "hosts" (https://github.com/docker/docker/issues/8681) and "devices" (https://github.com/docker/docker/pull/8826) changes are still a proposal, I think the creators those proposals should discuss a consistent UI before they're implemented to keep things consistent.

For the *existing* commands I understand that most frequently used commands are deliberately left short for convenience, perhaps "long" versions (`docker images rm`?) could be added to make the UI more consistent.

(This is a follow up on a comment I placed here https://github.com/docker/docker/pull/8484#issuecomment-60827328)

**update**
Added the "hosts" proposal to the list and sorted the list a bit

### LINKED ISSUE #13509: Proposal: Add a 'docker image' command
Right now docker's mixing of container and image commands leads to a bit of a usability problem. For example, `docker inspect` doesn't know if we're talking about containers or images, so we check for both. Which can lead to conflicts since images and containers can have the same name, and partial IDs.  Solutions like #13187 are ok but not really ideal.  We have similar issues with 'docker rm', but in that case we created 'docker rmi', which works, but feels hacky.

It would be better if there was a dedicated `docker image` command so that its clear we're talking about images and not containers.  It also allows us to add new image-specific commands (and container-specific commands at the top level) w/o the possibility of conflict.

While, people may naturally then ask about a `docker container` command, and that is certainly possible, but I'm not suggesting we do that right now.  I think keeping all `docker verb` commands acting on containers makes sense because docker's main focus is on containers and the extra typing involved in writing `docker container verb` would IMO be a UX hit.

I've create a branch for people to play with to see how it feels:  https://github.com/duglin/docker/tree/ImageCmd
All `docker image verb` commands just call the appropriate `docker verb` command so we can avoid duplication of logic/code.  However it would require more code than what I show in the branch in order to finish this work and to give a better UX, in particular:
- make the help for each image related command show both the `docker verb` and `docker image verb` help text in both flows
- will need to add code so that the error messages show the correct command based on what the user used.
- will need to leverage #13187 for `docker image inspect` or some other code if #13187 doesn't go in

I'm sure there's more, but this branch does seem to show that the basic idea does seem to work, but before I go much further I'd like to know if we'd like to head this direction.

To be clear, I'm not suggesting we remove the old `docker verb` commands at this time, we can't break people.  I think this path is consistent with the proposals I've heard to add other top-level `docker noun` commands (like `docker volume` and `docker network` ...).

ping @crosbymichael @shykes

### PR DESCRIPTION
**Pull request by dnephin**
- Fixes #13509
- Fixes #8829
- Fixes #8756

- create a new `docker image` and `docker container` command
- add the image and container commands as subcommands of the new command groups with a few commands renamed:
  - `images` -> `image ls` (with aliases)
  - `rmi` -> `image rm` (with aliases)
  - `ps` -> `container ls` (with aliases)
- add `image inspect` and `container inspect` to match the other command groups
- add 3-letter allises for all commands groups
  - `docker con`
  - `docker img`
  - `docker srv`
  - `docker net`
  - `docker vol`

### CODE REVIEW

**Comment by icecrime** [comment]
Absolutely LGTM on design, :tada:, :100:, etc.

My only nit, but I won't fight for it, is that I'm skeptical about the need for short aliases (`docker con`, `docker vol`, etc.).

**Comment by nathanleclaire** [comment]
> `docker con`

image

**Comment by MHBauer** [comment]
Not big on the short aliases, but I like the thought for everything else.

**Comment by vdemeester** [comment]
Not a huge fan of short aliases but other than that, just like @icecrime, absolutely LGTM on design :sunglasses: :tada:

**Comment by dnephin** [comment]
My rational for the short aliases is this:

It's going to be difficult to convince people to use `docker container rm` instead of `docker rm`. It's just much longer to type.  Adding the short aliases make the situation a lot better because it's just 3 extra characters, instead of 9!

If we don't make it easy to use the new canonical commands we're never going to be able to remove them.

**Comment by cpuguy83** [comment]
You think we'll be able to remove something like `docker rm` or `docker run`? I doubt that.

**Comment by dnephin** [comment]
`docker run` no, we aren't even hiding it, it's staying as a permanent alias.

The rest, not any time soon. Maybe in a year or two if we're good about converting all the docs to the new form, and communicating the new canonical way of doing things.

**Comment by stevvooe** [comment]
I love this, but if we do it, we should take this chance to fix the output of inspect to default to  be pretty and composite. This was a HUGE point of contention during the implementation of swarm services. Making `inspect` output text composed of multiple API calls will address this problem and let us really grow the usability of `inspect`, while addressing long standing issues.

File: api/client/image/cmd.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	cmd.AddCommand(
+		NewBuildCommand(dockerCli),
+		NewHistoryCommand(dockerCli),
+		NewImportCommand(dockerCli),
+		NewLoadCommand(dockerCli),
+		NewPullCommand(dockerCli),
+		NewPushCommand(dockerCli),
+		NewSaveCommand(dockerCli),
+		NewSearchCommand(dockerCli),
```
Comment (stevvooe) [inline api/client/image/cmd.go]: I don't think we should carry over search to `docker image`. We aren't searching the local image store, we are searching remote registries. Future extension to search won't happen under `docker image`.

File: api/client/image/cmd.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		Aliases: []string{"img"},
+		Args:    cli.NoArgs,
+		Run: func(cmd *cobra.Command, args []string) {
+			fmt.Fprintf(dockerCli.Err(), "\n"+cmd.UsageString())
+		},
+	}
+	cmd.AddCommand(
+		NewBuildCommand(dockerCli),
+		NewHistoryCommand(dockerCli),
+		NewImportCommand(dockerCli),
```
Comment (stevvooe) [inline api/client/image/cmd.go]: This is a container operation, if I remember correctly.

Comment (stevvooe) [inline api/client/image/cmd.go]: Nevermind. It is `docker export` that is for containers.

File: api/client/image/cmd.go (hunk 1 above)
Comment (icecrime) [inline api/client/image/cmd.go]: That's a good point. Maybe it belongs with `login` and `logout` under another top-level command?

**Comment by OJFord** [comment]
> My rational for the short aliases is this:
> It's going to be difficult to convince people to use `docker container rm` instead of `docker rm`. It's just much longer to type.

Realistically people won't do that though, they'll type `docker c<tab> rm`. As long as the completion is the full "container" (i.e. "con" isn't an entry) I'm not bothered, but for this reason I don't think it's necessary.

File: api/client/image/cmd.go (hunk 1 above)
Comment (dnephin) [inline api/client/image/cmd.go]: Makes sense. I'll move the search command file to the correct package, and move it back to the top level.

**Comment by dnephin** [comment]
>  we should take this chance to fix the output of inspect to default to be pretty and composite

I'm not sure that we can change `inspect`, but we could definitely introduce a new command that defaults to pretty print, and hide `inspect` from help. Do you know if there is an issue with a proposal for this change? I'd like to keep that discussion and change separate from this PR.

**Comment by nathanleclaire** [comment]
> Realistically people won't do that though, they'll type docker c<tab> rm. As long as the completion is the full "container" (i.e. "con" isn't an entry) it wouldn't bother me, but for this reason I don't think it's necessary.

My 2c -- I'm not really a fan of completion on subcommands, since it requires that you install and maintain additional completion scripts instead of being something that works out of the box.  In `docker container`'s case as well, keep in mind that it's competing with `docker commit` and others for the same completion, so by the time you've typed enough characters to differentiate `container` for completion, you've already typed `docker con` anyway.

I wouldn't really like to see `docker ps` go away pretty much ever since it is such deeply ingrained habit for me, is extremely short and frequently invoked and I don't really mind having porcelain around the "true" commands (git style, i.e. "every command is just a wrapper around rebase").  But I understand the motivation.

I'd definitely like to see more elaboration around where we draw the lines here, e.g., why would it be OK for `docker service scale foo=3` to alias to `docker service update --replicas 3 foo` (porcelain wrapper like mentioned above) but not OK for `docker ps` to alias to `docker container ls`.

File: api/client/image/cmd.go (hunk 1 above)
Comment (stevvooe) [inline api/client/image/cmd.go]: There may be a `docker registry` command, but that is a whole 'notha can o wyrms!

**Comment by stevvooe** [comment]
> I'm not sure that we can change inspect, but we could definitely introduce a new command that defaults to pretty print, and hide inspect from help. Do you know if there is an issue with a proposal for that change? I'd like to keep that discussion and change separate from this PR.

Not that I know of. File away or ping me again and I'll cook something up.

I agree that this shouldn't be a part of this PR, if we frame it that way.

**Comment by dnephin** [comment]
This is now rebased, and no longer hides the commands. It only adds the new canonical commands.

**Comment by dnephin** [comment]
rebased and green again

@icecrime what do you think?

(the commands are no longer hidden)

**Comment by icecrime** [comment]
@dnephin Thanks, I think this is less disruptive for the user this way. The only unfortunate (but I believe unavoidable) thing is those two very similar commands:

```
  image       Manage Docker images
  images      List images
```

1) Rather than removing the command hiding entirely, I would make it dependent on some environment variable perhaps? Please ignore the choice of the variable name, but for example:

```
diff --git a/cli/command/commands/commands.go b/cli/command/commands/commands.go
index 6d103d6..f972719 100644
--- a/cli/command/commands/commands.go
+++ b/cli/command/commands/commands.go
@@ -1,6 +1,8 @@
 package commands
 
 import (
+       "os"
+
        "github.com/docker/docker/cli/command"
        "github.com/docker/docker/cli/command/container"
        "github.com/docker/docker/cli/command/image"
@@ -73,6 +75,9 @@ func AddCommands(cmd *cobra.Command, dockerCli *command.DockerCli) {
 }
 
 func hide(cmd *cobra.Command) *cobra.Command {
+       if v := os.Getenv("DOCKER_HIDE_LEGACY_COMMANDS"); v == "" {
+               return cmd
+       }
        cmdCopy := *cmd
        cmdCopy.Hidden = true
        cmdCopy.Aliases = []string{}
```

This would give the following:

```
# docker

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Commands:
  attach      Attach to a running container
  build       Build an image from a Dockerfile
  commit      Create a new image from a container's changes
  container   Manage Docker containers
  cp          Copy files/folders between a container and the local filesystem
  create      Create a new container
  diff        Inspect changes on a container's filesystem
  events      Get real time events from the server
  exec        Run a command in a running container
  export      Export a container's filesystem as a tar archive
  history     Show the history of an image
  image       Manage Docker images
  images      List images
  import      Import the contents from a tarball to create a filesystem image
  info        Display system-wide information
  inspect     Return low-level information on a container, image or task
  kill        Kill one or more running containers
  load        Load an image from a tar archive or STDIN
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  logs        Fetch the logs of a container
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  pause       Pause all processes within one or more containers
  port        List port mappings or a specific mapping for the container
  ps          List containers
  pull        Pull an image or a repository from a registry
  push        Push an image or a repository to a registry
  rename      Rename a container
  restart     Restart one or more containers
  rm          Remove one or more containers
  rmi         Remove one or more images
  run         Run a command in a new container
  save        Save one or more images to a tar archive (streamed to STDOUT by default)
  search      Search the Docker Hub for images
  service     Manage Docker services
  start       Start one or more stopped containers
  stats       Display a live stream of container(s) resource usage statistics
  stop        Stop one or more running containers
  swarm       Manage Docker Swarm
  tag         Tag an image into a repository
  top         Display the running processes of a container
  unpause     Unpause all processes within one or more containers
  update      Update configuration of one or more containers
  version     Show the Docker version information
  volume      Manage Docker volumes
  wait        Block until one or more containers stop, then print their exit codes

Run 'docker COMMAND --help' for more information on a command.
# DOCKER_HIDE_LEGACY_COMMANDS=1 docker

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Commands:
  build       Build an image from a Dockerfile
  container   Manage Docker containers
  events      Get real time events from the server
  image       Manage Docker images
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  run         Run a command in a new container
  search      Search the Docker Hub for images
  service     Manage Docker services
  swarm       Manage Docker Swarm
  version     Show the Docker version information
  volume      Manage Docker volumes

Run 'docker COMMAND --help' for more information on a command.
```

2) In the help output, can we group together the "proper" ones from the legacy? Again, please ignore the choice of working, but for example:

```
# DOCKER_HIDE_LEGACY_COMMANDS=1 docker

Usage:  docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/root/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/root/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/root/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/root/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Operations:
  build       Build an image from a Dockerfile
  events      Get real time events from the server
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  run         Run a command in a new container
  search      Search the Docker Hub for images
  version     Show the Docker version information
```

**Comment by dnephin** [comment]
Sounds good. I'll make those changes

**Comment by dnephin** [comment]
I've made those changes, here's what it looks like by default.

```
Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/home/daniel/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/home/daniel/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/home/daniel/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/home/daniel/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Commands:
  attach      Attach to a running container
  build       Build an image from a Dockerfile
  commit      Create a new image from a container's changes
  cp          Copy files/folders between a container and the local filesystem
  create      Create a new container
  diff        Inspect changes on a container's filesystem
  events      Get real time events from the server
  exec        Run a command in a running container
  export      Export a container's filesystem as a tar archive
  history     Show the history of an image
  images      List images
  import      Import the contents from a tarball to create a filesystem image
  info        Display system-wide information
  inspect     Return low-level information on a container, image or task
  kill        Kill one or more running containers
  load        Load an image from a tar archive or STDIN
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  logs        Fetch the logs of a container
  pause       Pause all processes within one or more containers
  port        List port mappings or a specific mapping for the container
  ps          List containers
  pull        Pull an image or a repository from a registry
  push        Push an image or a repository to a registry
  rename      Rename a container
  restart     Restart one or more containers
  rm          Remove one or more containers
  rmi         Remove one or more images
  run         Run a command in a new container
  save        Save one or more images to a tar archive (streamed to STDOUT by default)
  search      Search the Docker Hub for images
  start       Start one or more stopped containers
  stats       Display a live stream of container(s) resource usage statistics
  stop        Stop one or more running containers
  tag         Tag an image into a repository
  top         Display the running processes of a container
  unpause     Unpause all processes within one or more containers
  update      Update configuration of one or more containers
  version     Show the Docker version information
  wait        Block until one or more containers stop, then print their exit codes

Run 'docker COMMAND --help' for more information on a command.
```

With hidden commands enabled by environment variable
```

Usage:	docker COMMAND

A self-sufficient runtime for containers.

Options:
      --config string      Location of client config files (default "/home/daniel/.docker")
  -D, --debug              Enable debug mode
      --help               Print usage
  -H, --host value         Daemon socket(s) to connect to (default [])
  -l, --log-level string   Set the logging level (default "info")
      --tls                Use TLS; implied by --tlsverify
      --tlscacert string   Trust certs signed only by this CA (default "/home/daniel/.docker/ca.pem")
      --tlscert string     Path to TLS certificate file (default "/home/daniel/.docker/cert.pem")
      --tlskey string      Path to TLS key file (default "/home/daniel/.docker/key.pem")
      --tlsverify          Use TLS and verify the remote
  -v, --version            Print version information and quit

Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes

Commands:
  build       Build an image from a Dockerfile
  events      Get real time events from the server
  info        Display system-wide information
  login       Log in to a Docker registry.
  logout      Log out from a Docker registry.
  run         Run a command in a new container
  search      Search the Docker Hub for images
  version     Show the Docker version information

Run 'docker COMMAND --help' for more information on a command.
```

**Comment by duglin** [comment]
I'm sure someone better at UX, or docs, that me could come up with something really smart (I can't right now), but let's brainstorm on this section:
```
Management Commands:
  container   Manage Docker containers
  image       Manage Docker images
  network     Manage Docker networks
  node        Manage Docker Swarm nodes
  service     Manage Docker services
  swarm       Manage Docker Swarm
  volume      Manage Docker volumes
```
The text on the right seems kind of obvious and therefore verbose.

Maybe even just list them, such as:
```
Resource Management Commands:
   container, image, network, node, service, swarm
```
I dunno....

**Comment by vdemeester** [comment]
LGTM 🐸
@dnephin needs a rebase 👼

**Comment by thaJeztah** [comment]
I agree with @duglin's comment as well.

Also (following the earlier discussion); I think we should remove the three-letter aliases for now (`docker con`, `docker img` etc.). It looks like we're not sure they're really needed, and in case of doubt, I suggest to *not* do it. It's way easier to add them later, then to deprecate.

If you feel strongly about the three-letter aliases, perhaps open a separate PR, so that this PR won't get blocked because of that. (Sorry for my late reply - blame my vacation :smile:)

**Comment by thaJeztah** [comment]
Oh, other than that (but I need to give this a more in-depth look); I love this PR!

**Comment by dnephin** [comment]
Hmm, I don't get any conflicts when I rebase with `master`.  I'll make those changes.

**Comment by dnephin** [comment]
Ok, I've removed the short aliases (I'll submit as a follow up PR for more discussion), and I've removed the word "Docker" from the usage test of the management commands as suggested.

I don't like the idea of listing them on a single line. I think they aren't noticeable enough that way, and they are going to look weird with text wrapping.

**Comment by icecrime** [comment]
LGTM again.

**Comment by mlaventure** [comment]
LGTM (assuming it's :green_heart:)

**Comment by thaJeztah** [comment]
Just discussed with @dnephin that we're gonna do docs (and completion scripts) in a follow-up PR. I'll open a tracking issue

**Comment by thaJeztah** [comment]
And, it's green!

**Comment by vieux** [comment]
LGTM !

**Comment by icecrime** [comment]
:clap:

**Comment by gittycat** [comment]
Is there any chance of differentiating between Legacy|old|standalone Swarm and "Swarm Mode" at least in the doc. For instance:

```
Management Commands:
  ...
  node        Manage Docker Swarm mode nodes
  service     Manage Docker services
  swarm       Manage Docker legacy Swarm
  ...
```

The old vs new swarm is a recurring area of misunderstanding over at Stackoverflow and elsewhere. The "swarm mode" should really have been named something else in my opinion. Sorry to bring this up here but since we're stuck with "Swarm" I would love the doc to at least acknowledge that it's dealing with two different technologies and use some naming to help distinguish them.

**Comment by dnephin** [comment]
@gittycat  `docker swarm` is not the legacy swarm, it's the new "swarm mode". Only `swarm` is the old swarm, which is not available through this CLI.

**Comment by stevvooe** [comment]
@gittycat Rule of thumb: if there is a `docker` in front of your command, you're probably using Swarm-mode.

I usually refer to them as "Docker Services" for clarification.

**Comment by albers** [comment]
ping @sdurrheimer there's a lot of work waiting for you here :cold_sweat:, PTAL

**Comment by sdurrheimer** [comment]
@albers I see 😲, thx for the ping.

**Comment by lpalgarvio** [comment]
kudos!

i was actually using custom make scripts to do just this with almost equal commands!!!

thanks for standardizing in 1.13! saved me from opening a ticket =)

**Comment by icetech233** [comment]
整那么多乱七八糟的，我日你m

**Pull request merged**